src/fairseq2/recipe/optim.py

```diff
@@ -5,17 +5,242 @@
+
+                break
+
+    output: list[dict[str, object]] = []
+
+    for idx, group in enumerate(groups):
+        if not group.empty:
+            if group.is_fallback:
+                continue
+
+            log.warning("Optimizer parameter group {} is empty.", idx)
+        elif log.is_enabled_for_info():
+            s = ", ".join(sorted(n for n in group.param_names))
+
+            log.info("Optimizer Parameter Group {}: {}", idx, s)
+
+        kwargs = group.get_kwargs()
+
+        output.append(kwargs)
+
+    return output
+
+
+@final
+class ParameterGroup:
+    """
+    Represents an optimizer parameter group, used as input to the
+    :func:`prepare_parameter_groups` function.
+    """
+
+    def __init__(
+        self, name_patterns: str | Sequence[str], config: object, fields: Sequence[str]
+    ) -> None:
+        """
+        :param name_patterns: The regular expression(s) used to select the
+            parameters that belong to this group.
+        :param config: An opaque object -*typically a dataclass*- that holds the
+            configuration of this group.
+        :param fields: The names of the configuration fields that ``config``
+            holds. Any field that has a non-default value will be passed as a
+            group ``kwarg`` to the optimizer.
+        """
+        if isinstance(name_patterns, str):
+            name_patterns = [name_patterns]
+
+        self._name_patterns = name_patterns
+        self._config = config
+        self._fields = fields
+        self._params: list[Parameter] = []
+        self._param_names: list[str] = []
+
+    def is_match(self, name: str) -> bool:
+        """
+        Returns ``True`` if ``name`` matches one of the name patterns of this
+        group.
+
+        :param name: The name to check.
+        """
+        return any(name == p or re.match(p, name) for p in self._name_patterns)
+
+    def add_parameter(self, param_name: str, param: Parameter) -> None:
+        """
+        Adds ``param`` to this group.
+
+        :param param_name: The name of the parameter, used for logging and error
+            reporting purposes.
+        :param param: The parameter tensor.
+        """
+        self._params.append(param)
+
+        self._param_names.append(param_name)
+
+    def get_kwargs(self) -> dict[str, object]:
+        """Returns the group ``kwargs`` to be passed to the optimizer."""
+        kwargs: dict[str, object] = {"params": self._params}
+
+        for field in self._fields:
+            value = getattr(self._config, field)
+            if value != "default":
+                kwargs[field] = value
+
+        return kwargs
+
+    @property
+    def param_names(self) -> Iterable[str]:
+        """Gets the names of the parameters belonging to this group."""
+        return self._param_names
+
+    @property
+    def empty(self) -> bool:
+        """Gets whether this group has no parameters."""
+        return bool(self._params)
```

Comment:
This seems to be a logical error to me? Should be:
```suggestion
        return not bool(self._params)
```